electron.wrapElectronHook: init (#487711)
NixOS/nixpkgs: doc/hooks/electron-wrap.section.md

      -c.electronDist=electron-dist
      -c.electronVersion=${electron.version}
  '';

  installPhase = ''
    runHook preInstall

    mkdir -p "$out/share"

    cp -r dist/resources $out/share/electron-program

    runHook postInstall
  '';

  meta = {
    mainProgram = "electronProject";
    description = "electron project";
  };
})
```
:::

## Variables controlling `electronWrapHook` {#electron-wrap-hook-variables}

### `electronWrapHook` Exclusive Variables {#electron-wrap-hook-exclusive-variables}

#### `electronAppAsar` {#electron-wrap-hook-asar}

Controls the `*.asar` file that is wrapped.
Usually does not need to be set as the hook will be able to find the file automatically.

#### `electronWrapProgram` {#electron-wrap-hook-program}

Controls the name of the wrapper program.
This will be output to `{!outputBin}/bin/$name`, which is either `$bin` or `$out` in that order of precedence.
Often does not need to be set, and `meta.mainProgram` should be set instead.
Only set this if the Electron program is not the *main* program of the package.

#### `electronWrapperArgs` {#electron-wrap-hook-args}

Controls the arguments to `makeWrapper`.

Default flags:
- `--inherit-argv0`

> - `--set-default ELECTRON_IS_DEV 0`

Author: Cool, I'll remove it.

# electron.electronWrapHook {#electron-wrap-hook}

> Hook for wrapping Electron-based packages.

Reviewer: Maybe the manual section could explain why one would want to use the wrapper and what it does?

Author: Why? This is not really ever done. The reason is always the same: so that Nixpkgs maintainers can change how builds across Nixpkgs are done by editing one file.

## Examples {#electron-wrap-hook-example}

:::{.example #electron-wrap-hook-example-snippet}

# Using `electronWrapHook`

Note that how Electron packages are built can vary widely between packages.

```nix
{
  stdenv,
  fetchFromGitHub,
  fetchPnpmDeps,
  pnpmConfigHook,
  pnpm,
  electron,
}:
buildNpmPackage (finalAttrs: {
  pname = "electron-project";
  version = "1.0";

  src = fetchFromGitHub {
    owner = "JaneElectron";
    repo = "ElectronProject";
    tag = finalAttrs.version;
    hash = "...";
  };

  npmDeps = null;
  pnpmDeps = fetchPnpmDeps {
    inherit (finalAttrs) pname version src;
    inherit pnpm;
    fetcherVersion = 3;
    hash = "...";
  };

  npmConfigHook = pnpmConfigHook;

  # Must be in buildInputs since the Electron is the
  # Electron running on the host machine
  buildInputs = [
    electron.electronWrapHook
  ];

  env = {
    ELECTRON_SKIP_BINARY_DOWNLOAD = 1;

>     npm_config_nodedir = electron.headers;

Reviewer 2: Pretty important to note: `npmConfigHook` has `export npm_config_nodedir=@node_headers@` in it, which would override `env.npm_config_nodedir`, so packages using the npm hooks usually add `--ignore-scripts` to `npmRebuildFlags` and then rebuild manually or leave it to electron-builder to maybe do the rebuild.

However, an important thing to note is that `npm_config_*` will stop working in future nodejs versions (where `*` is a non-standard config value, eg. `nodedir`).
This probably means node-gyp will introduce a new env-var we can override.

  };

  postBuild = ''

>     cp -r ${electron.dis} electron-dist

Author: Good catch

# electron.electronWrapHook {#electron-wrap-hook}

Hook for wrapping Electron-based packages.

## Examples {#electron-wrap-hook-example}

:::{.example #electron-wrap-hook-example-snippet}

# Using `electronWrapHook`

Note that how Electron packages are built can vary widely between packages.

```nix
{
  stdenv,
  fetchFromGitHub,
  fetchPnpmDeps,
  pnpmConfigHook,
  pnpm,
  electron,
}:
buildNpmPackage (finalAttrs: {
  pname = "electron-project";
  version = "1.0";

  src = fetchFromGitHub {
    owner = "JaneElectron";
    repo = "ElectronProject";
    tag = finalAttrs.version;
    hash = "...";
  };

  npmDeps = null;
  pnpmDeps = fetchPnpmDeps {
    inherit (finalAttrs) pname version src;
    inherit pnpm;
    fetcherVersion = 3;
    hash = "...";
  };

  npmConfigHook = pnpmConfigHook;

  # Must be in buildInputs since the Electron is the
  # Electron running on the host machine
  buildInputs = [

>     electron.electronWrapHook

Reviewer 3: 1. Why not as a separate electronWrapHook attribute? For pnpm we use pnpmConfigHook and in your future pnpmBuildHook as separate atributes [`8da59b8` (#487046)](https://github.com/NixOS/nixpkgs/pull/487046/changes/8da59b8f176624862ce4abf710ef51048157d377)
2. Same question and reasoning for why in buildInputs and not in nativeBuildInputs

Reviewer 2: You're probably thinking of something like this, right?
https://github.com/NixOS/nixpkgs/pull/487711#issuecomment-4041809050

Reviewer 3: Yeah

Author: 1. Why should it be separate? This logically connects the Electron that is being used to the hook. We are not, and cannot, execute Electron in this hook from PATH so there's not a great way of searching for the Electron to wrap without making the user provide the Electron via a function. 
2. This makes it so that splicing selects the correct Electron. There are other frameworks that do their host hooks this way.

Author: I think it makes sense to pnpm as you can just slap down the pnpm you want in `nativeBuildInputs` and it just works because it is added to PATH. For this we cannot do such a thing. So the alternative is something like:
```
electronWrapHook {
  electron = electron_38;
};
```
which just seems ugly to me.

For the hook in `buildInputs` vs. in `nativeBuildInputs`, I think it's just easier conceptually to place it in `buildInputs` since this is a host->host hook, and so splicing takes care of the Electron selected.

Reviewer 4: `pnpm` is used at build-time and `electron` is used at runtime, so they're not comparable. I was initially wary of this approach, but the alternatives are worse. If you want electron from `PATH`, you'd need `targetPkgs.electron` or such - most packages would be liable to get it wrong, and we need proper education on dependencies and cross-compilation before we start enforcing stuff like that.

or if `electron-pacakger` is used and `asar=true` is not passed. In that case the path that is wrapped will be
a directory, not an archive file.

#### `electronWrapperName` {#electron-wrap-hook-name}

Controls the name of the wrapper program.
This will be output to `{!outputBin}/bin/$name`, which is either `$bin` or `$out` in that order of precedence.
Often does not need to be set, and `meta.mainProgram` should be set instead.

Only set this if the Electron program is not the *main* program of the package, as by default it reads the value set in
`meta.mainProgram`.

#### `electronWrapperArgs` {#electron-wrap-hook-args}

Controls the arguments to `makeWrapper`.

Default flags:
- The `*.asar` file to wrap
- `--set-default ELECTRON_FORCE_IS_PACKAGED 1`

If for example this Electron program requires arguments from other toolkits, those flags can be appended like so:

```nix
{
  #...
  dontWrapGApps = true;
  preInstall = ''
    electronWrapperArgs+=("''${gappsWrapperArgs[@]}")
  '';
  #...
}
```
#### `dontWrapElectron` {#electron-wrap-hook-dont}

Control whether the hook is ran or not. Defaults to `false`.

> If set to false, the hook will not do any detection.

Reviewer 2: ```suggestion
If set to true, the hook will not do any detection.
```

Author: Thanks. It used to be opposite, a `doElectronWrap`, but then I changed it since most hook configs are `dont*`

> # electron.wrapElectronHook {#electron-wrap-hook}

Reviewer 2: ```suggestion
# electron.wrapElectronHook {#wrap-electron-hook}
```

Author: This is internal to documentation and does not matter.

Hook for wrapping Electron-based packages.

> ## Examples {#electron-wrap-hook-example}

Reviewer 2: ```suggestion
## Examples {#wrap-electron-hook-example}
```

:::{.example #electron-wrap-hook-example-snippet}

> # Using `electronWrapHook`

Reviewer 2: ```suggestion
# Using `wrapElectronHook`
```

Note that how Electron packages are built can vary widely between packages.

```nix
{
  stdenv,
  fetchFromGitHub,
  fetchPnpmDeps,
  pnpmConfigHook,
  pnpm,

>   electron,
> }:
> buildNpmPackage (finalAttrs: {
>   pname = "electron-project";
>   version = "1.0";
> 
>   src = fetchFromGitHub {
>     owner = "JaneElectron";
>     repo = "ElectronProject";
>     tag = finalAttrs.version;
>     hash = "...";
>   };
> 
>   npmDeps = null;
>   pnpmDeps = fetchPnpmDeps {
>     inherit (finalAttrs) pname version src;
>     inherit pnpm;
>     fetcherVersion = 3;
>     hash = "...";
>   };
> 
>   npmConfigHook = pnpmConfigHook;
> 
>   # Must be in buildInputs since the Electron is the
>   # Electron running on the host machine
>   buildInputs = [
>     electron.wrapElectronHook
>   ];
> 
>   env = {
>     ELECTRON_SKIP_BINARY_DOWNLOAD = 1;
>   };
> 
>   postBuild = ''
>     cp -r ${electron.dist} electron-dist
>     chmod -R u+w electron-dist
> 
>     npm_config_nodedir=${electron.headers} \
>     npm exec electron-builder -- \
>       --config electron-builder.json \
>       --dir \
>       -c.electronDist=electron-dist
>       -c.electronVersion=${electron.version}
>   '';
> 
>   installPhase = ''
>     runHook preInstall
> 
>     mkdir -p "$out/share"
> 
>     cp -r dist/resources $out/share/electron-program
> 
>     runHook postInstall
>   '';
> 
>   meta = {
>     mainProgram = "electronProject";
>     description = "electron project";
>   };
> })

Reviewer 2: No need to complicate it by using pnpm.

```suggestion
{
  stdenv,
  buildNpmPackage,
  fetchFromGitHub,
  electron,
}:
buildNpmPackage (finalAttrs: {
  pname = "electron-project";
  version = "1.0";

  src = fetchFromGitHub {
    owner = "JaneElectron";
    repo = "ElectronProject";
    tag = finalAttrs.version;
    hash = "...";
  };

  npmDepsHash = "...";

  # Must be in buildInputs since the Electron is the
  # Electron running on the host machine
  buildInputs = [
    electron.wrapElectronHook
  ];

  env = {
    ELECTRON_SKIP_BINARY_DOWNLOAD = 1;
  };

  postBuild = ''
    cp -r ${electron.dist} electron-dist
    chmod -R u+w electron-dist

    export npm_config_nodedir=${electron.headers}
    npm exec electron-builder -- \
      --config electron-builder.json \
      --dir \
      -c.electronDist=electron-dist
      -c.electronVersion=${electron.version}
  '';

  installPhase = ''
    runHook preInstall

    mkdir -p "$out/share"

    cp -r dist/resources $out/share/electron-program

    runHook postInstall
  '';

  meta = {
    mainProgram = "ElectronProject";
    /* ... */
  };
})
```

Author: There's no need to simplify it either, people write more than npm program.

```
:::

## Variables controlling `electronWrapHook` {#electron-wrap-hook-variables}

### `electronWrapHook` Exclusive Variables {#electron-wrap-hook-exclusive-variables}

> #### `wrapElectronPath` {#electron-wrap-hook-path}

Reviewer 2: ```suggestion
## Variables controlling `wrapElectronHook` {#wrap-electron-hook-variables}

### `wrapElectronHook` Exclusive Variables {#wrap-electron-hook-exclusive-variables}

#### `wrapElectronPath` {#wrap-electron-hook-path}
```

Controls the path that is wrapped by Electron.
Most often this is a `*.asar` archive that is automatically found by the hook.
But in some cases this will need to be manually provided:

1. If there are multiple `*.asar` archives in one package
2. If a `*.asar` archive was not generated

Archives are not always generated, often in cases when `asar=false` is passed to `electron-builder` (if used),
or if `electron-pacakger` is used and `asar=true` is not passed. In that case the path that is wrapped will be
a directory, not an archive file.

#### `electronWrapperName` {#electron-wrap-hook-name}

Controls the name of the wrapper program.
This will be output to `{!outputBin}/bin/$name`, which is either `$bin` or `$out` in that order of precedence.

> Often does not need to be set, and `meta.mainProgram` should be set instead.

Reviewer 2: Should probably explicitly mention that it actually uses `meta.mainProgram` by default

Author: Good point.

Only set this if the Electron program is not the *main* program of the package.

> #### `electronWrapperArgs` {#electron-wrap-hook-args}

Reviewer 2: ```suggestion
#### `electronWrapperArgs` {#wrap-electron-hook-args}
```

Controls the arguments to `makeWrapper`.

Default flags:
- The `*.asar` file to wrap

> - `--set-default ELECTRON_FORCE_IS_PACKAGED 1`

Reviewer 2: Is there a way to add flags without using bash? Or will it just completely override the defaults if you set `electronWrapperArgs` in the derivation attrs?

Author: You can append flags in bash as long as it is before the hook. There is an example right below this snippet showing so.

Author: Oh wait I misread.

Author: The defaults cannot be overridden, you can only append to them.

If for example this Electron program requires arguments from other toolkits, those flags can be appended like so:

```nix
{
  #...
  dontWrapGApps = true;
  preInstall = ''
    electronWrapperArgs+=("''${gappsWrapperArgs[@]}")
  '';
  #...
}
```

> #### `dontWrapElectron` {#electron-wrap-hook-dont}

Reviewer 2: ```suggestion
#### `dontWrapElectron` {#wrap-electron-hook-dont}
```

Control whether the hook is ran or not. Defaults to `true`.

> If set to false, the hook will not do any detection.

Reviewer 2: This is backwards...

Author: Good catch

This is useful for manually wrapping paths using the wrapper manually.

The wrapper as two required arguments, and then optional arguments.

1. The path to wrap, generally a `*.asar` file
2. The path to place the wrapper.
3. All other argument will be appended to the `makeWrapper` call.

```nix
{

>   dontElectronWrap = true;

Reviewer 2: ```suggestion
  dontWrapElectron = true;
```

Author: Good catch

  postInstall = ''
    mkdir -p "$out/bin"

>     electronWrap "$out/share/myApp/app.asar" "$out/bin/myApp" --set SOME_ENV 1

Reviewer 2: ```suggestion
    wrapElectron "$out/share/myApp/app.asar" "$out/bin/myApp" --set SOME_ENV 1
```

  '';
}
```

#### `electronWrapPackaged` {#electron-wrap-hook-is-packaged}

Controls whether the Electron pacakged is told it is in a production package.
This variable is defauled to `true`, so if the attribute is not set, it is assumed that

> the app is told to run in production mode.

Reviewer 2: Other than the typos, this section should probably clear up what "production" means. Also, the variable name is pretty bad, maybe `wrapElectronForcePackaged`?

```suggestion
#### `wrapElectronForcePackaged` {#wrap-electron-hook-force-packaged}

Controls whether the Electron package is told it is in a production package.
This variable is defaulted to `true`, so if the attribute is not set, it is assumed that
the app is told to run in production mode.
```

Author: Yeah good enough

If the app breaks under these conditions, it can be disabled by setting this attribute to false.

```nix
{

>   electronWrapPackaged = false;

Reviewer 2: ```suggestion
  wrapElectronForcePackaged = false;
```

}
```

The reason that some application break under this is because it changes how Electron passes the elements of `process.argv` to the application.

When `ELECTRON_FORCE_IS_PACKAGED` is false, Electron runs in "dev mode", meaning that arvg looks like:
```js
[
  "/nix/store/aaaaaaaaaaaaaaaaaaaaaaaaaaaaa-electron/bin/electron",
  "/nix/store/aaaaaaaaaaaaaaaaaaaaaaaaaaaaa1-coolapp/share/coolapp/resources/app.asar",
  "--version"
]
```
Notice that the second argument is the path to the `asar`. When the `ELECTRON_FORCE_IS_PACKAGED` is true, it is the following:
```js
[
  "/nix/store/aaaaaaaaaaaaaaaaaaaaaaaaaaaaa-electron/bin/electron",
  "--version"

> ]

Reviewer 2: AFAICT this is not true.

`process.argv` is always `[path/to/electron, path/to/app.asar, first-arg] when running via the original electron instead of the copied one.

Setting `ELECTRON_FORCE_IS_PACKAGED` just forces `app.isPackaged` to be true, it doesn't change `process.argv` layout.

Apps that take arguments do something like this to calculate what their args are:
`const args = process.argv.slice(app.isPackaged ? 1 : 2)`
However this doesn't work if we force `app.isPackaged`, but `process.argv` is still the "dev" layout.

This is one of the things that can be broken by setting `ELECTRON_FORCE_IS_PACKAGED=1`.

But then if it supposedly breaks many things, why do apps set it?

E.g. some apps try to search different directories for bundled files when `isPackaged` is true, since `electron-builder` moves files around.

I will bring more examples later, but I'm still experimenting with something currently.

Author: Oh, what I wrote is just how I interpreted your above message. So I see now, we force it to be packaged but it is not.

Author: My guess is that:
1. The devs in electron/electron#4690 recommend using a library to paper over the differences, maybe some take `ELECTRON_FORCE_IS_PACKAGED` into account
2. Most people don't actually pass args to Electron programs

so the breakage isn't seen often

Author: * `yargs`: Doesn't account for it
* `command-line-args`: Doesn't account for it

So that hypothesis is wrong

Author: I have a funny idea that may fix this

Author: https://github.com/RossSmyth/nixpkgs/tree/electron-wrapper-symlink

Ok my idea was based upon your comment that "renaming the Electron binary is what signals to Electron it is packaged." So I tried basically doing that.

1. Put a symlink to Electron in the output path of the derivation
2. Make the wrapper use that symlink as the "electron"

This works in that it does think it is packaged, but for for example, Signal fails to start because it cannot find some file in the asar:
```
Unhandled Promise Rejection: Error: ENOENT, _locales/en/messages.json not found in /nix/store/vgkns0bk1d4mp5mj6cpicx2w80hw497v-signal-desktop-8.9.1/share/signal-desktop/app.asar
    at createError (node:electron/js2c/node_init:2:2073)
    at readFileFromArchiveSync (node:electron/js2c/node_init:2:10211)
    at t.readFileSync (node:electron/js2c/node_init:2:10995)
    at Zv (bundles:///main.js:84:17097)
    at iy (bundles:///main.js:84:18633)
    at EventEmitter.<anonymous> (bundles:///main.js:84:47605)
[global_errors] Unhandled Promise Rejection: Error: ENOENT, _locales/en/messages.json not found in /nix/store/vgkns0bk1d4mp5mj6cpicx2w80hw497v-signal-desktop-8.9.1/share/signal-desktop/app.asar
    at createError (node:electron/js2c/node_init:2:2073)
    at readFileFromArchiveSync (node:electron/js2c/node_init:2:10211)
    at t.readFileSync (node:electron/js2c/node_init:2:10995)
    at Zv (bundles:///main.js:84:17097)
    at iy (bundles:///main.js:84:18633)
    at EventEmitter.<anonymous> (bundles:///main.js:84:47605)
```

which I did confirm (via `asar list result/share/signal-desktop/app.asar`) that said file is not in the .asar.
https://github.com/signalapp/Signal-Desktop/blob/bd36b19a848ec5639e827894377ee847e2c2c8b3/_locales/en/messages.json

Building and running rocketchat-desktop and mattermost-desktop works fine though, it just seems like those files aren't packaged in the Signal asar for some reason. 

But it seems like this doesn't actually change anything. Looks like the symlink is dereferenced, so it is executed via the electron store path, not the symlink. Even if `--inherit-argv0` is added back. 

So in the end it is a failure, nothing changed or it got worse.

```
In "production mode," Electron drop the `asar` path. Not all apps handle this correctly. The proper way to handle this is the [following code](https://github.com/electron/electron/issues/4690#issuecomment-422617581):
```js
if (app.isPackaged) {
  process.argv.unshift(null);
}

> ```

Reviewer 2: This section is very wrong.
`ELECTRON_FORCE_IS_PACKAGED` changes nothing, but the fact that `app.isPackaged` is true.
Even the "proper way" is wrong I think.

Author: Yeah this was written when I had misunderstood what you had written.
